presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregationIntoTableScan.java

```diff
@@ -0,0 +1,214 @@
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.tableScan;
+
+public class PushAggregationIntoTableScan
+        implements Rule<AggregationNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .with(source().matching(tableScan().capturedAs(TABLE_SCAN)));
+
+    private final Metadata metadata;
+
+    public PushAggregationIntoTableScan(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode node, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+        Map<String, ColumnHandle> assignments = tableScan.getAssignments()
+                .entrySet().stream()
+                .collect(toImmutableMap(entry -> entry.getKey().getName(), Entry::getValue));
+
+        List<Entry<Symbol, AggregationNode.Aggregation>> aggregations = node.getAggregations()
+                .entrySet()
+                .stream()
+                .collect(toImmutableList());
+
+        List<AggregateFunction> aggregateFunctions = toAggregateFunction(aggregations
+                .stream()
+                .map(Entry::getValue)
+                .collect(toImmutableList()));
+        List<Symbol> aggregationOutputSymbols = aggregations.stream()
+                .map(Entry::getKey)
+                .collect(toImmutableList());
+
+        if (aggregateFunctions.isEmpty()) {
+            return Result.empty();
+        }
+
+        GroupingSetDescriptor groupingSets = node.getGroupingSets();
+
+        if (groupingSets.getGroupingSetCount() > 1) {
+            return Result.empty();
+        }
+
+        List<ColumnHandle> groupByColumns = groupingSets.getGroupingKeys().stream()
+                .map(groupByColumn -> assignments.get(groupByColumn.getName()))
+                .collect(toImmutableList());
+
+        Optional<AggregationPushdownResult<TableHandle>> aggregationPushdownResult = metadata.applyAggregation(
+                context.getSession(),
+                tableScan.getTable(),
+                aggregateFunctions,
+                assignments,
+                groupByColumns.isEmpty() ? ImmutableList.of() : ImmutableList.of(groupByColumns));
+
+        if (!aggregationPushdownResult.isPresent()) {
+            return Result.empty();
+        }
+
+        AggregationPushdownResult<TableHandle> result = aggregationPushdownResult.get();
+
+        TableHandle handle = result.getHandle();
+
+        // The new scan outputs and assignments are union of existing and the ones returned by connector pushdown.
+        ImmutableList.Builder<Symbol> newScanOutputs = new ImmutableList.Builder<>();
+        newScanOutputs.addAll(tableScan.getOutputSymbols());
```

Comment:
This is incorrect. The new scan outputs should be the symbols associated with grouping columns plus the symbols associated with aggregations.

Reply:
comment removed.